### PR TITLE
Fix click.prompt() printing a single space to stdout when err=True.

### LINKED ISSUE #2018: click.prompt(..., err=True) produces a single space on stdout
Using the following script

```python
import click

click.prompt("asd", err=True)
```

and running it like

```
python test.py 1> stdout 2> stderr
```

The file `stderr` should receive all output from the script, and `stdout` should remain empty. However we can see that the file `stdout` receives a single space for each round that `click.prompt()` does (i.e. when entering nothing, it keeps re-asking).

This is unexpected and should not happen.

I believe [this comment](https://github.com/pallets/click/blob/main/src/click/termui.py#L141) is related to this issue.

Environment:

- Python version: 3.7.10
- Click version: 8.0.1

### PR DESCRIPTION
**Pull request by nils-werner**
When using `click.prompt(..., err=True)`, the underlying code sends a single space to `stdout`, which may confuse programs who are trying to digest the output. This PR fixes that by sending the space to whatever output stream was chosen instead. For `err=True`, the space is sent to `sys.stdout`.

Note that I am not sure how to test this behaviour, that's why this PR is a WIP. Any suggestions are welcome.

- fixes #2018

Checklist:

- [ ] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by davidism** [comment]
The comment above the code you've changed calls out using `stdout` for the space specifically.

> Echo a space to stdout to work around an issue where readline causes backspace to clear the whole line.

This was from https://github.com/pallets/click/issues/665 and https://github.com/pallets/click/pull/1836. Please confirm that this does not cause the issue with readline again before I can consider merging this.

**Comment by nils-werner** [comment]
I checked and can confirm that it does not regress issue #665.

I checked running

```python
import click, readline

click.prompt("Input")
```

and

```python
import click, readline

click.prompt("Input", err=True)
```

and hitting backspace. The prompt wasn't deleted in either case.

File: src/click/termui.py
Comment (davidism) [inline src/click/termui.py:56]: This is not correct. Click purposefully uses `sys.stdout`, not the original `__stdout__`, in order to allow overriding `stdout`, such as in tests.

File: src/click/termui.py
Comment (davidism) [inline src/click/termui.py:72]: This `try/finally` isn't needed here, `getpass` takes a `stream` parameter.

**Comment by davidism** [comment]
It seems like while your issue report regarded printing a space to stdout, what you were really reporting was that Python's `input` and `getpass` always print to stdout. While your patching approach does work, it's something Click deliberately avoids doing outside of tests. I'm not comfortable with merging this type of change.

> may confuse programs who are trying to digest the output

It seems like it might be easier / safer to adjust how that program is understanding the output, such as ignoring empty space, rather than patching `sys.stdout` for `input`.

I do want to allow the developer to override what things like `prompt` do, so that might help with this use case eventually.

**Comment by GaretJax** [comment]
> It seems like it might be easier / safer to adjust how that program is understanding the output, such as ignoring empty space, rather than patching sys.stdout for input.

@davidism that sadly not always possible, as the program digesting the output is not always under our control.

**Comment by davidism** [comment]
@nils-werner were you planning to address my comments and review?

**Comment by nils-werner** [comment]
I was under the impression that you had rejected my proposed changes.

If not, could you repeat what you wanted to see changed?

File: src/click/termui.py
Comment (nils-werner) [reply src/click/termui.py:56]: I am not sure I understand. The purpose of `sys.__stdout__` et al is to be the [original value of `sys.stdout`, and the Python docs recommend to use it to reset them to that original value](https://docs.python.org/3/library/sys.html#sys.__stdin__).

Comment (davidism) [reply src/click/termui.py:56]: Tools like pytest and Click's CliRunner already override `sys.stdout`. We would want to use those, not the originals.

Comment (nils-werner) [reply src/click/termui.py:56]: OK, understood.

**Comment by davidism** [comment]
OK, I think this change makes sense given the other existing capabilities, but we're going to need to document that using `prompt` with `stderr` is not thread (or coroutine) safe because it doesn't lock the change in global state. If you're looking to make a contribution to Python, maybe open an issue that `input` should take a stream to show output on, like `getpass`.

**Comment by davidism** [comment]
I'm surprised that readline continues to work. I don't see anything about it knowing what stream to re-output on, it should always use stdout from what I can tell.

**Comment by davidism** [comment]
I think I've got a different solution. Changing the approach in #1836:

```diff
-            echo(text.rstrip(" "), nl=False, err=err)
+            echo(text, nl=False, err=err)
             # Echo a space to stdout to work around an issue where
             # readline causes backspace to clear the whole line.
-            return f(" ")
+            return f(" \b")
```

Don't strip the space off the prompt, so it's written to the stream as in 7.x. However, instead of sending a space to `input`, send a space *and backspace*. This prints the entire prompt to stderr, still triggers whatever in readline preserves the prompt, and doesn't actually include the extra space or backspace character in either stream.

**Review by jab — CHANGES_REQUESTED** [review]
Before we moved forward with an approach like this (that temporarily reassigns the shared global resource sys.stdout, which could maybe pull the carpet out from under some programs), first it seems worth having a discussion with the Python maintainers about this issue, and whether it should be fixed in Python itself (e.g. by adding support for `input(..., stream=sys.stderr)`, or whether they have some better recommendation than this).

In the meantime, certainly seems worth documenting the current behavior of the space being written to stdout. Would you like to submit that as a separate PR in the meantime, @nils-werner?

(While I was looking at this quickly, I also caught a possible issue with a couple of the type hints, though that's very minor.)

**Pull request closed without merging**
